### DISCUSSION TITLE
Textured Geometry on HUD Camera

### DISCUSSION THREAD

**Discussion by nebsar** [pelicanmapping/osgearth#2033] [Support Q&A] [answered]
This may be an OpenSceneGraph question. But I would like to ask here as well. I am trying to show an image on the HUD. I am using almost the same code used in osgEarth Controls.cpp. You may see the code below:

	osg::ref_ptr<osg::Image> image = osgDB::readRefImageFile("resources/images/image.png");

	osg::Geometry* g = new osg::Geometry();
	g->setUseVertexBufferObjects(true);
	g->setDataVariance(osg::Object::DYNAMIC);

	osg::ref_ptr<osg::Texture2D> texture = new osg::Texture2D(image.get());

	osg::Vec3f _renderPos(-image->s()/2, image->t()/2, 0.0);
	osg::Vec2f _renderSize(image->s(),image->t());

	float rx = osg::round(_renderPos.x());
	float ry = osg::round(_renderPos.y());
	float vph = image->t();

	osg::Vec3Array* verts = new osg::Vec3Array(6);
	g->setVertexArray(verts);

	(*verts)[0].set(rx, vph - ry, 0);
	(*verts)[1].set(rx, vph - ry - _renderSize.y(), 0);
	(*verts)[2].set(rx + _renderSize.x(), vph - ry - _renderSize.y(), 0);
	(*verts)[3].set((*verts)[2]);
	(*verts)[4].set(rx + _renderSize.x(), vph - ry, 0);
	(*verts)[5].set((*verts)[0]);

	g->addPrimitiveSet(new osg::DrawArrays(GL_TRIANGLES, 0, 6));

	osg::Vec4Array* c = new osg::Vec4Array(osg::Array::BIND_OVERALL, 1);
	(*c)[0] = osg::Vec4f(1, 1, 1, 1);
	g->setColorArray(c);

	bool flip = false;
	if (image.valid())
		flip = (image->getOrigin() == osg::Image::TOP_LEFT);
	else if (texture->getImage(0))
		flip = (texture->getImage(0)->getOrigin() == osg::Image::TOP_LEFT);

	osg::Vec2Array* t = new osg::Vec2Array(6);

	(*t)[0].set(0, flip ? 0 : 1);
	(*t)[1].set(0, flip ? 1 : 0);
	(*t)[2].set(1, flip ? 1 : 0);
	(*t)[3].set((*t)[2]);
	(*t)[4].set(1, flip ? 0 : 1);
	(*t)[5].set((*t)[0]);

	g->setTexCoordArray(0, t);

	g->getOrCreateStateSet()->setTextureAttributeAndModes(0, texture.get(), osg::StateAttribute::ON);
	g->getOrCreateStateSet()->addUniform(new osg::Uniform("oe_Controls_tex", (int)0));

        osg::ref_ptr<osg::MatrixTransform> transformation = new osg::MatrixTransform;
        transformation->addChild(g);
        _geode->addChild(transfomation);
       _hudCamera->addChild(_geode);

But the image looks like a white rectangle, instead of a small png icon.

![image](https://user-images.githubusercontent.com/5169314/160649939-2d9d32f8-d3df-46fd-a1e3-88ee12e5ffb2.png)

I want to show the below image in the center.

![F-35A_Cyan](https://user-images.githubusercontent.com/5169314/160650157-2039b72b-fc8f-4a5d-a671-d437243c1dcc.png)

**Comment by gwaldron** [comment, marked as answer]
This line implies that you are using a shader -- can we see the shader code?
```c++
g->getOrCreateStateSet()->addUniform(new osg::Uniform("oe_Controls_tex", (int)0));
```

**Reply by nebsar** [reply]
Hi Glenn,

I am not much familiar with the shaders. I think I forgot to include the shader code in my code as well. When I included the shader code, it worked well. Thank you again.

![image](https://user-images.githubusercontent.com/5169314/160702046-c38aaf67-0c88-46b9-aafb-5599c0a24c9b.png)
